src/client/src/client/client.rs

```diff
@@ -37,15 +38,15 @@ pub(crate) trait SerialFrame: Sync + Send + Display {
     }
 
     /// send and receive
-    async fn send_receive<R>(&mut self, request: R) -> Result<R::Response, FlvSocketError>
+    async fn send_receive<R>(&self, request: R) -> Result<R::Response, FlvSocketError>
```

Comment:
What's use case for this?

Also, only send (Sink) need to be mutable.   There is already `InnerExclusiveFlvSink` for that.

Reply:
The original thing I was thinking about wanting to do was cache different sockets to different replica leaders. For example, in the producer, we use `spu_pool.create_serial_socket` on every record send. I was thinking it would be good to have like a `HashMap<ReplicaKey, VersionedSerialSocket>` to cache the sockets so we wouldn't need to open them over and over. But doing that is complicated when the sockets require `&mut` access to do anything.

To be clear, I'm not convinced that caching these sockets will _actually_ get us any benefit, but I noticed that changing this `&mut` to a `&` made the API cleaner all around. Currently the only way to achieve this is to wrap the inner socket in a Mutex because the sockets we are using are not thread-safe. If we were to switch to thread-safe sockets in the future, this API change would still benefit from it.

Reply:
SerialFrame is designed for used in a single task  not multi-threaded.  Whole purpose of `MultiplexorSocket` to allow multiple concurrent operation to a single socket.  It doesn't make sense to share a same serial SerialFrame in different thread.  So we should defer until if we actually hit the use case